src/coreclr/jit/gentree.cpp

```diff
@@ -21967,3 +21967,27 @@ bool GenTree::IsInvariant() const
     GenTree* lclVarTree = nullptr;
     return OperIsConst() || Compiler::impIsAddressInLocal(this, &lclVarTree);
 }
+
+//------------------------------------------------------------------------
+// IsNeverNegative: returns true if the given tree is known to be never
+//      negative such as: casts to unsiged, UDIV, UMOD, Array.Length
+//
+// Return Value:
+//      true if the given tree is known to be never negative
+//
+bool GenTree::IsNeverNegative() const
+{
+    if (IsUnsigned() || OperIs(GT_UMOD, GT_UDIV, GT_ARR_LENGTH))
```

Comment:
1) `IsUnsigned` != `never negative`: `CAST_UN(-1 -> uint)` is `-1`.
2) `GT_UDIV` can be negative too: `uint.MaxValue / 1`.

I think this should call `IntegralRange::ForNode(this, compiler).IsPositive()`. Ideally the range would be used directly, but if that's not as good because it doesn't handle constants yet, it seems fine to have this wrapper too.

(But if it turns out to catch more cases, please leave a `TODO-Casts: extend IntegralRange to handle constants and delete this method.` for me to take a look at it later.